In [268]:
#Import libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline,FeatureUnion
from sklearn.model_selection import GridSearchCV
from sklearn.feature_selection import VarianceThreshold
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

# 1. Import the dataset and ensure that it loaded properly.

In [269]:
df_loan = pd.read_csv("Loan_Train.csv")

In [270]:
df_loan.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

# 2a. Drop the column “Load_ID.”

In [271]:
df_loan = df_loan.drop("Loan_ID",axis=1)

In [272]:
df_loan

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
0      Male      No          0      Graduate            No             5849   
1      Male     Yes          1      Graduate            No             4583   
2      Male     Yes          0      Graduate           Yes             3000   
3      Male     Yes          0  Not Graduate            No             2583   
4      Male      No          0      Graduate            No             6000   
..      ...     ...        ...           ...           ...              ...   
609  Female      No          0      Graduate            No             2900   
610    Male     Yes         3+      Graduate            No             4106   
611    Male     Yes          1      Graduate            No             8072   
612    Male     Yes          2      Graduate            No             7583   
613  Female      No          0      Graduate           Yes             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                  0.0         NaN             360.0             1.0   
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

    Property_Area Loan_Status  
0           Urban           Y  
1           Rural           N  
2           Urban           Y  
3           Urban           Y  
4           Urban           Y  
..            ...         ...  
609         Rural           Y  
610         Rural           Y  
611         Urban           Y  
612         Urban           Y  
613     Semiurban           N  

[614 rows x 12 columns]

# 2b. Drop any rows with missing data

In [273]:
df_loan = df_loan.dropna()

In [274]:
df_loan

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
1      Male     Yes          1      Graduate            No             4583   
2      Male     Yes          0      Graduate           Yes             3000   
3      Male     Yes          0  Not Graduate            No             2583   
4      Male      No          0      Graduate            No             6000   
5      Male     Yes          2      Graduate           Yes             5417   
..      ...     ...        ...           ...           ...              ...   
609  Female      No          0      Graduate            No             2900   
610    Male     Yes         3+      Graduate            No             4106   
611    Male     Yes          1      Graduate            No             8072   
612    Male     Yes          2      Graduate            No             7583   
613  Female      No          0      Graduate           Yes             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
5               4196.0       267.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

    Property_Area Loan_Status  
1           Rural           N  
2           Urban           Y  
3           Urban           Y  
4           Urban           Y  
5           Urban           Y  
..            ...         ...  
609         Rural           Y  
610         Rural           Y  
611         Urban           Y  
612         Urban           Y  
613     Semiurban           N  

[480 rows x 12 columns]

# 2c. Convert the categorical features into dummy variables.

In [275]:
df_loan = pd.get_dummies(df_loan,
                     columns = df_loan.select_dtypes(include='object').keys())

In [276]:
df_loan.head()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   

   Credit_History  Gender_Female  Gender_Male  Married_No  Married_Yes  \
1             1.0              0            1           0            1   
2             1.0              0            1           0            1   
3             1.0              0            1           0            1   
4             1.0              0            1           1            0   
5             1.0              0            1           0            1   

   Dependents_0  ...  Dependents_3+  Education_Graduate  \
1             0  ...              0                   1   
2             1  ...              0                   1   
3             1  ...              0                   0   
4             1  ...              0                   1   
5             0  ...              0                   1   

   Education_Not Graduate  Self_Employed_No  Self_Employed_Yes  \
1                       0                 1                  0   
2                       0                 0                  1   
3                       1                 1                  0   
4                       0                 1                  0   
5                       0                 0                  1   

   Property_Area_Rural  Property_Area_Semiurban  Property_Area_Urban  \
1                    1                        0                    0   
2                    0                        0                    1   
3                    0                        0                    1   
4                    0                        0                    1   
5                    0                        0                    1   

   Loan_Status_N  Loan_Status_Y  
1              1              0  
2              0              1  
3              0              1  
4              0              1  
5              0              1  

[5 rows x 22 columns]

# 3. Split the data into a training and test set, where the “Loan_Status” column is the target.

In [277]:
x_data = df_loan.loc[:, ~df_loan.columns.isin(['Loan_Status_Y', 'Loan_Status_N'])].values
y_data = df_loan[['Loan_Status_Y']].values

In [278]:
x_data

array([[4.583e+03, 1.508e+03, 1.280e+02, ..., 1.000e+00, 0.000e+00,
        0.000e+00],
       [3.000e+03, 0.000e+00, 6.600e+01, ..., 0.000e+00, 0.000e+00,
        1.000e+00],
       [2.583e+03, 2.358e+03, 1.200e+02, ..., 0.000e+00, 0.000e+00,
        1.000e+00],
       ...,
       [8.072e+03, 2.400e+02, 2.530e+02, ..., 0.000e+00, 0.000e+00,
        1.000e+00],
       [7.583e+03, 0.000e+00, 1.870e+02, ..., 0.000e+00, 0.000e+00,
        1.000e+00],
       [4.583e+03, 0.000e+00, 1.330e+02, ..., 0.000e+00, 1.000e+00,
        0.000e+00]])

In [279]:
y_data

array([[0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
    

In [280]:
#Training and tes datasets split
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data ,test_size = 0.2, shuffle=False)

# 4. Create a pipeline with a min-max scaler and a KNN classifier

In [281]:
#Create a pipeline

pipe= Pipeline([
("scaler", MinMaxScaler()),
("classifier", KNeighborsClassifier())
])

# 5. Fit a default KNN classifier to the data with this pipeline. Report the model accuracy on the test set.

In [282]:
#Fitting Pipeline Model
pipe.fit(x_train,y_train)

/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


Pipeline(steps=[('scaler', MinMaxScaler()),
                ('classifier', KNeighborsClassifier())])

In [283]:
#score the knn model on the testing data
pipe.score(x_test,y_test)

0.7291666666666666

# 6. Create a search space for your KNN classifier where your “n_neighbors” parameter varies from 1 to 10.

In [284]:
#Create space of candidate values
k_range = list(range(1, 10))
param_grid = dict(n_neighbors=k_range)

# 7. Fit a grid search with your pipeline, search space, and 5-fold cross-validation to find the best value for the “n_neighbors” parameter.

In [290]:
knn_gscv= GridSearchCV(knn, param_grid, cv=5, scoring='accuracy', return_train_score=False,verbose=1)

In [291]:
knn_gscv.fit(x_train,y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits


/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversi

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9]},
             scoring='accuracy', verbose=1)

In [292]:
#check top performing n_neighbors value
knn_gscv.best_params_

{'n_neighbors': 9}

# 8. Find the accuracy of the grid search best model on the test set

In [293]:
knn_predict = knn_gscv.predict(x_test)

In [294]:
knn_predict

array([1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 0], dtype=uint8)

In [295]:
# Stores the optimum model in best_pipe
best_pipe = knn_gscv.best_estimator_
print(best_pipe)

KNeighborsClassifier(n_neighbors=9)


In [296]:
#check mean score for the top performing value of n_neighbors
knn_gscv.best_score_

0.6769993164730008

In [297]:
cross_val_score(knn_gscv, x_data, y_data, cv=5)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
Fitting 5 folds for each of 9 candidates, totalling 45 fits


/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversi

/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversi

Fitting 5 folds for each of 9 candidates, totalling 45 fits
Fitting 5 folds for each of 9 candidates, totalling 45 fits


/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversi

/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversi

Fitting 5 folds for each of 9 candidates, totalling 45 fits


/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversi

array([0.67708333, 0.6875    , 0.67708333, 0.65625   , 0.66666667])

In [298]:
knn_accuracy = accuracy_score(y_test,knn_predict)
knn_accuracy

0.6770833333333334

In [299]:
# Best performing model and its corresponding hyperparameters
knn_gscv.best_params_

{'n_neighbors': 9}

In [300]:
# ROC-AUC score for the best model
knn_gscv.best_score_

0.6769993164730008

# 9. Now, repeat steps 6 and 7 with the same pipeline, but expand your search space to include logistic regression and random forest models with the hyperparameter values in section 12.3 of the Machine Learning with Python Cookbook

In [301]:
log_model = LogisticRegression(solver='lbfgs', max_iter=1000)

In [302]:
#Create dictionary with candidate learning algos and their hyperparameters
search_space = [{"classifier":[log_model],
                "classifier__penalty": ['l1','l2'],
                "classifier__C": np.logspace(0,4,10)},
               {"classifier":[RandomForestClassifier()],
                "classifier__n_estimators": [10,100,1000],
                "classifier__max_features": [1,2,3]}]

In [303]:
#Create grid search
gridsearch = GridSearchCV(pipe,search_space,cv=5,verbose=0)

In [304]:
#Fit grid search
best_model = gridsearch.fit(x_train,y_train.ravel())

/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: 
50 fits failed out of a total of 145.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
50 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/pipeline.py", line 394, in fit
    self._final_estimator.fit(Xt, y, **fit_params_last_step)
  File "/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages

In [305]:
best_model

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', MinMaxScaler()),
                                       ('classifier', KNeighborsClassifier())]),
             param_grid=[{'classifier': [LogisticRegression(C=59.94842503189409,
                                                            max_iter=1000)],
                          'classifier__C': array([1.00000000e+00, 2.78255940e+00, 7.74263683e+00, 2.15443469e+01,
       5.99484250e+01, 1.66810054e+02, 4.64158883e+02, 1.29154967e+03,
       3.59381366e+03, 1.00000000e+04]),
                          'classifier__penalty': ['l1', 'l2']},
                         {'classifier': [RandomForestClassifier()],
                          'classifier__max_features': [1, 2, 3],
                          'classifier__n_estimators': [10, 100, 1000]}])

# What are the best model and hyperparameters found in the grid search? Find the accuracy of this model on the test set.

In [306]:
# Best performing model and its corresponding hyperparameters
best_model.best_params_

{'classifier': LogisticRegression(C=59.94842503189409, max_iter=1000),
 'classifier__C': 59.94842503189409,
 'classifier__penalty': 'l2'}

In [307]:
# ROC-AUC score for the best model
best_model.best_score_

0.7943950786056049

In [308]:
# View best model
best_model.best_estimator_.get_params()["classifier"]

LogisticRegression(C=59.94842503189409, max_iter=1000)

In [309]:
best_model_predict = best_model.predict(x_test)

In [310]:
best_model_predict

array([1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1,
       0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 0], dtype=uint8)

In [311]:
best_model_accuracy = accuracy_score(y_test,best_model_predict)
best_model_accuracy

0.7916666666666666

In [312]:
cross_val_score(best_model, x_data, y_data.ravel(), cv=5)

/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: 
50 fits failed out of a total of 145.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
50 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages/sklearn/pipeline.py", line 394, in fit
    self._final_estimator.fit(Xt, y, **fit_params_last_step)
  File "/Users/somashekarvayuvegula/opt/anaconda3/lib/python3.9/site-packages

array([0.80208333, 0.78125   , 0.77083333, 0.85416667, 0.82291667])

# Summarize your results.

We have analyzed Grid search with two scenarios:
    1. Using KNN Classifier
    2. Uisng LogisticRegression and RandomForestClassifier
    
Upon the prediction and accuracy test, Accuracy score for the first case is 67.7% and second case is 79.1%. Model looks better when we have used LogisticRegression and RandomForestClassifier.

Best score for first case is 67.7% and second case is 79.4%

Second case scenario is the better compared to the first one.